crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
     }
 }
 
+#[derive(Debug)]
+enum StringQuoteChar {
+    Single,
+    Double,
+}
+
+impl TryFrom<char> for StringQuoteChar {
+    type Error = String;
+
+    fn try_from(c: char) -> Result<Self, Self::Error> {
+        match c {
+            '\'' => Ok(Self::Single),
+            '"' => Ok(Self::Double),
+            _ => Err(format!("Invalid string quote character: {c}")),
+        }
+    }
+}
+
+impl StringQuoteChar {
+    const fn as_char(&self) -> char {
+        match self {
+            Self::Single => '\'',
+            Self::Double => '"',
+        }
+    }
+}
+
+#[derive(Debug)]
+enum StringQuoteSize {
+    Single,
+    Triple,
+}
+
+#[derive(Debug)]
+struct FStringContext {
+    quote_char: StringQuoteChar,
+    quote_size: StringQuoteSize,
+    parentheses_count: u32,
+    format_spec_count: u32,
```

Comment:
Can we document these fields and why tracking them in the state is necessary? E.g. why do we need to track both `parenthesis_count` and `format_spec_count` (Should it be `format_spec_depth`?). What are the different possible states?

Reply:
There are 3 reasons to track these fields:
1. To check if we're in a f-string expression `f"{<here>}"`
2. To check if we're in a format spec within a f-string expression `f"{foo:<here>}"`
3. To check if the `:` is to start a format spec or is it part of, for example, named expression `:=`:
	* For `f"{x:=1}"`, the colon is to indicate the format spec start
	* For `f"{(x:=1)}"`, the colon is part of `:=` named expression because the colon is not at the same level of parentheses. Another example is `f"{x,{y:=1}}"` where the colon is part of `:=` named expression

From [PEP 701: How to produce these new tokens](https://peps.python.org/pep-0701/#how-to-produce-these-new-tokens):
> 3. [..] This mode tokenizes as the “Regular Python tokenization” until a `:` or a `}` character is encountered with the same level of nesting as the opening bracket token that was pushed when we enter the f-string part. [..]

With this explanation, do you think `_depth` is a better suffix?